SPARK-16420: Ensure compression streams are closed.

rdblue opened the pull request:
## What changes were proposed in this pull request?

This uses the try/finally pattern to ensure streams are closed after use. `UnsafeShuffleWriter` wasn't closing compression streams, causing them to leak resources until garbage collected. This was causing a problem with codecs that use off-heap memory.
## How was this patch tested?

Current tests are sufficient. This should not change behavior.

rdblue committed:
SPARK-16420: Ensure compression streams are closed.

SparkQA commented:
**[Test build #61916 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61916/consoleFull)** for PR 14093 at commit [`14ba621`](https://github.com/apache/spark/commit/14ba62197e57ace8940ce71f5a6e5608b6765661).

srowen commented on core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java:
```diff
@@ -349,12 +349,19 @@ void forceSorterToSpill() throws IOException {
+              ByteStreams.copy(partitionInputStream, mergedFileOutputStream);
+              innerThrewException = false;
+            } finally {
+              Closeables.close(partitionInputStream, innerThrewException);
```
Not that it matters much, but what about also using the `Utils.tryWithSafeFinally` pattern here for some consistency?

rdblue replied:
This is the usual equivalent for Java code.

srowen replied:
Oh duh it's Java. I have made that mistake about 10 times now

SparkQA commented:
**[Test build #61916 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61916/consoleFull)** for PR 14093 at commit [`14ba621`](https://github.com/apache/spark/commit/14ba62197e57ace8940ce71f5a6e5608b6765661).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61916/
Test FAILed.

rdblue committed:
SPARK-16420: Do not foward close in LimitedInputStream.

This adds an option to LimitedInputStream to not close the underlying
stream when it is reused, which is the case in UnsafeShuffleWriter.

SparkQA commented:
**[Test build #61926 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61926/consoleFull)** for PR 14093 at commit [`601f934`](https://github.com/apache/spark/commit/601f934372922b3b68424d3ef5a3cc81fd0f4500).

SparkQA commented:
**[Test build #61926 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61926/consoleFull)** for PR 14093 at commit [`601f934`](https://github.com/apache/spark/commit/601f934372922b3b68424d3ef5a3cc81fd0f4500).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61926/
Test PASSed.

rxin commented on common/network-common/src/main/java/org/apache/spark/network/util/LimitedInputStream.java:
```diff
@@ -102,4 +118,10 @@ public LimitedInputStream(InputStream in, long limit) {
     left -= skipped;
     return skipped;
   }
+  @Override
```
add a blank line here?

rxin commented on core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java:
```diff
@@ -349,12 +349,19 @@ void forceSorterToSpill() throws IOException {
-            if (compressionCodec != null) {
-              partitionInputStream = compressionCodec.compressedInputStream(partitionInputStream);
+            InputStream partitionInputStream = null;
+            boolean innerThrewException = true;
```
this is a little bit strange - can we change it to do try catch and at the end of try we close with the flag false, and in catch we close with the flag true and then eliminate the variable?

rdblue replied:
This is the typical way to use `Closeables.close` in Java. The same pattern is used for the outer try/finally that is already there and is in [Guava's documentation|https://google.github.io/guava/releases/19.0/api/docs/com/google/common/io/Closeables.html#close(java.io.Closeable, boolean)].

I'm reluctant to change from the standard pattern for aesthetic reasons when `finally` is the correct control flow -- the stream should always be closed regardless of exceptions.

JoshRosen replied:
Yeah, I agree with @rdblue here; note that we use this `Closables.close` pattern elsewhere in Spark's Java code as well.

rxin commented:
cc @JoshRosen

rdblue committed:
SPARK-16420: Fix whitespace.

SparkQA commented:
**[Test build #61988 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61988/consoleFull)** for PR 14093 at commit [`661539f`](https://github.com/apache/spark/commit/661539f82ef924cfa4fe8a43f570c1a93dac816f).

JoshRosen commented:
LGTM pending Jenkins.

SparkQA commented:
**[Test build #61988 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61988/consoleFull)** for PR 14093 at commit [`661539f`](https://github.com/apache/spark/commit/661539f82ef924cfa4fe8a43f570c1a93dac816f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61988/
Test PASSed.

rxin commented:
Merging in master/2.0.
